safenode/src/transfers/mod.rs

```diff
@@ -0,0 +1,216 @@
+        Self {
+            our_name,
+            node_reward_key,
+            spend_queue: SpendQ::with_fee(STARTING_FEE),
+            storage: SpendStorage::new(),
+        }
+    }
+
+    /// Get Spend from local store.
+    pub(crate) async fn get(&self, address: DbcAddress) -> Result<SignedSpend> {
+        self.storage.get(address).await
+    }
+
+    /// Get the required fee for the specified spend priority.
+    pub(crate) fn get_required_fee(&self, dbc_id: DbcId, priority: SpendPriority) -> RequiredFee {
+        let amount = self.current_fee(priority);
+
+        debug!("Returned amount for priority {priority:?}: {amount}");
+
+        RequiredFee::new(Token::from_nano(amount), dbc_id, &self.node_reward_key)
+    }
+
+    /// Get the current fee for the specified spend priority.
+    fn current_fee(&self, priority: SpendPriority) -> u64 {
+        let spend_q_snapshot = self.spend_queue.snapshot();
+        let spend_q_stats = spend_q_snapshot.stats();
+        spend_q_stats.map_to_fee(priority)
+    }
+
+    /// Tries to add a double spend that was detected by the network.
+    pub(crate) async fn try_add_double(
+        &mut self,
+        a_spend: &SignedSpend,
+        b_spend: &SignedSpend,
+    ) -> Result<()> {
+        self.storage.try_add_double(a_spend, b_spend).await
+    }
+
+    /// Tries to add a new spend to the queue.
+    ///
+    /// All the provided data will be validated, and
+    /// if it is valid, the spend will be pushed onto the queue.
+    pub(crate) async fn try_add(
+        &mut self,
+        signed_spend: Box<SignedSpend>,
+        source_tx: Box<DbcTransaction>,
+        fee_ciphers: BTreeMap<XorName, FeeCiphers>,
+        parent_spends: BTreeSet<SignedSpend>,
+    ) -> Result<()> {
+        // 1. Validate the tx hash.
+        // Ensure that the provided src tx is the same as the
+        // one we have the hash of in the signed spend.
+        let provided_src_tx_hash = source_tx.hash();
+        let signed_src_tx_hash = signed_spend.src_tx_hash();
+
+        if provided_src_tx_hash != signed_src_tx_hash {
+            return Err(Error::SignedSrcTxHashDoesNotMatchProvidedSrcTxHash {
+                signed_src_tx_hash,
+                provided_src_tx_hash,
+            });
+        }
+
+        // 2. Try extract the fee paid for this spend, and validate it.
+        let paid_fee = self.validate_fee(source_tx.as_ref(), fee_ciphers)?;
+
+        // 3. Validate the spend itself.
+        self.storage.validate(signed_spend.as_ref()).await?;
+
+        // 4. Validate the parents of the spend.
+        // This also ensures that all parent's dst tx's are the same as the src tx of this spend.
+        validate_parent_spends(signed_spend.as_ref(), source_tx.as_ref(), parent_spends)?;
+
+        // This spend is valid and goes into the queue.
+        self.spend_queue.push(*signed_spend, paid_fee.as_nano());
+
+        // If the rate limit has elapsed..
+        if self.spend_queue.elapsed() {
```

Comment:
this means the `spend_queue` got checked only where there is another spend got pushed to the spend_queue?
which will result in:
1, some spend got processed far late
2, the last pushed spend never got processed

Reply:
Yeah, exactly. But that's totally fine now. And if we donn't see plenty of requests, the network is basically dead.

Reply:
but this doesn't make sense to rely one spend process on the next one.
as there will always be quiet period.
especially the spend request only sent to the close group?
which will make some nodes being quieter than other ?

Reply:
Yep, I thought of this as well. But it works. This is first iteration. Simplicity and get stuff up and running now.
As they say: "Make it work. Make it work as it should. Make it fast."